enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,195 @@
+
+An example of something that is NOT a critical alert:
+* MCO and/or related components are completely dead/crash looping.
+
+Even though the above is quite significant, there is no risk of immediate loss
+of the cluster.
+
+You might be thinking to yourself: "But my component is **super** important, and
+if it's not working, the user can't do X, which user definitely will want."  All
+that is probably true, but that doesn't make your alert "critical" it just makes
+it worthy of an alert.
+
+### Warning Alerts
+
+TL/DR: fix this soon, or some things won't work and upgrades will probably be
+blocked.
+
+If your alert does not meet the criteria in "Critical Alerts" above, it belongs
+to the warning level or lower.
+
+Most alerts probably fit in this category.  This level should represent degraded
+or unavailable services that do not jeopardize the immediate life of cluster.
+
+Timeline:  ~60 minutes
+
+60 minutes?  That seems high!  That's because it is high.  We want to reduce
+the noise.  We've done a lot to make clusters and operators auto-heal
+themselves.  The whole idea is that if a condition has persisted for more than
+60 minutes, it's unlikely to be resolved without intervention any time later.
+
+#### Example 1: All machine-config-daemons are crashlooping.
+
+That seems significant
+and it probably is, but nobody has to get out of bed to fix this **right now**.
+But, what if a machine dies, I won't be able to get a replacement?  Yeah, that
+is probably true, but how often does that happen?  Also, that's an entirely
+separate set of concerns.
+
+#### Example 2:  I have an operator that needs at least 3 of something to work.
+* With 2/3 replicas, warning alert after 60M.
+* With 1/3 replicas, warning alert after 10M.
+* With 0/3 replicas, warning alert after 5M.
+
+Q: How long should we wait until the above conditions rises to a 'critical'
+alert?
+
+A: It should never rise to the level of a critical alert, it's not critical.  If
+it was, this section would not apply.
+
+#### Example 3:  I have an operator that only needs 1 replica to function
+
+Well, if the cluster can upgrade with only 1 replica, and the the service is
+not unavailable despite other replicas being unavailable, this can probably
+be just an info-level alert.  We should have a "minor" alert, TBD.  But for now,
+"info" is the closest thing.
+
+#### Q: What if a particular condition will block upgrades?
+
+A: It's a warning level alert.
+
+
+### Documentation Required
+
+Every alert needs to have a URL to relevant documentation to resolve the issue.
+
+If there's not documentation, there shouldn't be an alert.
+
+Sometimes an alert's impact is not obvious.  We should state the impact of the
+alert.  Don't expect users to understand the context, we have too many moving
+pieces.
+
+
+### Open Questions [optional]
+
+Frankly, having just "Critical" and "Warning" is not helpful.
+
+We really need "Critical", "Major", "Minor" and "Info."  Can/Should we do this?
+
+### Test Plan
```

Comment:
I would actually love to see a test plan here, adding tests to e2e origin of openshift which have some best practices tested for alerts would be good here. Otherwise we are in the same spot in 6 months :) WDYT?

Reply:
I'm not opposed to a test plan, but I don't have one ATM.  I'm open to suggestion here.

Reply:
Take out a few compute machines.  Wait an hour.  Confirm that no critical alerts are firing.

Reply:
A few ideas:
* Ensure that all alerts have a `severity` label and comply with the list of allowed values. 
* Ensure that no critical alert gets added without a sign-off from teams that are on-call (e.g. OSD).
* Ensure that all alerts have mandatory annotations (e.g. `description` and `summary`).

Reply:
Do you actually need to test the alert condition duration or can this be sped up, eg, do I actually need to take out some machines to see if the alert would fire? If I do need to take them out, do I have to wait for the hour period? 

In terms of testing for alerts I was wondering if we could have some more static analysis style rules but perhaps that's a bit naive in at this level

Reply:
`promtool` has the ability to test alerts. Take a look at https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/ and example in https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/tests.yaml#L7-L32

Reply:
Also in the dynamic bucket, running a privileged DaemonSet on all cluster nodes that SIGKILLs a random process on its host every 5m.  With some kind of shared state to ensure only one of the control-plane pods fires each period; we don't want etcd losing quorum.  And then our existing check to make sure no criticals ever fire.

Reply:
And something else I'd like to see in a test plan is that Telemetry can squeeze at least one critical alarm out of the cluster when we try to terminate all its instances simultaneously at shutdown.  If some other, non-shutdown hiccup was able to quasi-simultaneously ask my hosts to gracefully shutdown, and nobody had turned off the Alertmanager syncs as part of a planned decommissioning process, I'd certainly want to get paged at midnight.

Reply:
I would like folks opinion if we need to resolve a test plan approach for alerts in the context of merging this enhancement.

Reply:
I would like a test plan, because policies that are not backed up by automated tests are hard to preserve.  Doesn't mean we need to have the tests implemented before we can start pushing towards new policy, but we should at least have an idea of what the tests will look like.

Reply:
New e2e tests in openshift/origin:

* Test validating that all alerting rules have a severity label matching the set of allowed values. 
* Test validating that all alerting rules have a summary and description annotations.
* Test validating that all runbook_url annotations point to an existing web page.

Reply:
@simonpasquier link to these tests?